### ISSUE TITLE
Socket mode is duplicating the messages

### ISSUE THREAD

**Issue by rcsrao** [slackapi/python-slack-events-api#93]
We have enabled the socket mode and when we send a message to the app, it is keep sending the same message,

'type': 'events_api', 'accepts_response_payload': False, 'retry_attempt': 0, 'retry_reason': ''
'type': 'events_api', 'accepts_response_payload': False, 'retry_attempt': 1, 'retry_reason': 'timeout'
'type': 'events_api', 'accepts_response_payload': False, 'retry_attempt': 2, 'retry_reason': 'timeout'
'type': 'events_api', 'accepts_response_payload': False, 'retry_attempt': 2, 'retry_reason': 'timeout'
'type': 'events_api', 'accepts_response_payload': False, 'retry_attempt': 3, 'retry_reason': 'timeout'
'type': 'events_api', 'accepts_response_payload': False, 'retry_attempt': 3, 'retry_reason': 'timeout'

How to address this? We are not using any SDK to handle the case instead interested in rest APIs.

**Comment by seratch** [comment]
@rcsrao This old library does not support Socket Mode and we don't have any plans to add the support in this project. Socket Mode is not a quite different way to communicate with the Slack server-side. It requires WebSocket connections, not an HTTP endpoint like this library supports.

If you use Socket Mode for your app, you can use Bolt for Python (or its underlying low-level SDK).

* https://slack.dev/bolt-python/concepts#socket-mode
* https://slack.dev/python-slack-sdk/socket-mode/index.html

If Socket Mode is not a right solution for you, you can turn it off at any time.

If everything is clear, would you mind closing this question?

**Comment by rcsrao** [comment]
@seratch We are not suing any SDK to establish the connection with socket and here it is how are making it,\

 ```
 self.url = requests.post('https://slack.com/api/apps.connections.open', headers=headers).json()['url']
 self.ws = websocket.WebSocketApp(self.url,
                                         on_message=self.on_message,
                                         on_error=self.on_error,
                                         on_close=self.on_close,
                                         on_open=self.on_open
                                         )
 self.ws.run_forever(http_proxy_host=self.proxy_host, http_proxy_port=self.proxy_port)
 ```
 
Problem is not with connection or receiving the events, but it keeps sending the same message and we would like to prevent that. Is there a way to send the ack response back to any end point for no retry or any other way to achieve the same?

**Comment by seratch** [comment]
@rcsrao If you implement a Socket Mode client from scratch, you need to implement not only the simple communications with the Slack server but also reconnecting logic and so on. For this reason, I don't recommend implementing it on your own but if you do so for some reason,

>Is there a way to send the ack response back to any end point for no retry or any other way to achieve the same?

You can do the same with what the SDK does. Also, reading [the document](https://api.slack.com/apis/connections/socket-implement#acknowledge) should be helpful.

Reference:
* https://github.com/slackapi/python-slack-sdk/blob/v3.5.1/slack_sdk/socket_mode/client.py#L87-L96
* https://github.com/slackapi/python-slack-sdk/blob/v3.5.1/slack_sdk/socket_mode/websocket_client/__init__.py#L206-L209

The `current_session` is a `WebSocketApp` instance.

**Comment by misscoded** [comment]
Hi @rcsrao! As we've provided an answer and have not heard back for some time, we're going to close this issue. We hope the information provided was helpful. 🙂

**Comment by coolblknerd** [comment]
Hey, so I was dealing with this issue while trying to create a slackbot using socketmode but it was in Go. What I ended up doing was using redis to capture the timestamp of the messages that were coming in. If the timestamp that came in matched the last message that was processed (indicating that it was a duplicate) I would discard the message. I kept running into this issue when I would make a `app_mention` call. Apparently, if you're also listening for `message` calls too, then those `app_mention` calls come in as `message` calls too, so as a result not only would it make two calls to whatever functions you have set-up, but if those functions are making other external calls then it can block the incoming websocket messages from getting processed. This causes a timeout on these messages which makes Slack think that the need to be retried. This sometimes caused functions to be called 5/6 times. It's a really hairy but that took some time to figure out. Hopefully this explanation helps you. If you want more details on what may be going on I'll be glad to help!
